src/librustc_passes/ast_validation.rs

```diff
@@ -58,9 +80,21 @@ impl<'a> AstValidator<'a> {
     }
 
     fn with_impl_trait(&mut self, outer: Option<Span>, f: impl FnOnce(&mut Self)) {
+        self.bound_context_stack.push(outer.map(|_| BoundContext::ImplTrait));
         let old = mem::replace(&mut self.outer_impl_trait, outer);
         f(self);
         self.outer_impl_trait = old;
+        self.bound_context_stack.pop();
+    }
+
+    fn with_bound_context(&mut self, ctx: Option<BoundContext>, f: impl FnOnce(&mut Self)) {
+        self.bound_context_stack.push(ctx);
+        f(self);
+        self.bound_context_stack.pop();
+    }
+
+    fn innermost_bound_context(&mut self) -> Option<BoundContext> {
+        self.bound_context_stack.iter().rev().find(|x| x.is_some()).copied().flatten()
```

Comment:
Instead of having a stack, you could have a single variable (like `outer_impl_trait`) and just not replace and swap back in case of None

Reply:
I did this so `with_bound_context` is now just like the other `with` functions. Kinda wish there were a more standardized way to do this, since this pattern is all over the place but often requires a copy-pasted `walk_x` equivalent.